src/RamCondition.h

```diff
@@ -31,138 +31,208 @@
 
 namespace souffle {
 
-/** abstract class condition */
+/**
+ * Abstract Class for RAM condition
+ */
 class RamCondition : public RamNode {
 public:
     RamCondition(RamNodeType type) : RamNode(type) {}
 
-    ~RamCondition() override = default;
-
-    /** get level of condition */
+    /** Get level */
     virtual size_t getLevel() = 0;
+
+    /** Create clone */
+    RamCondition* clone() const override = 0;
 };
 
-/** class for matching a string with a pattern */
+/**
+ * Conjunction
+ */
+// TODO: rename to RAMConjunction
 class RamAnd : public RamCondition {
-    /** left-hand side */
+protected:
+    /** Left-hand side of conjunction */
     std::unique_ptr<RamCondition> lhs;
 
-    /** right-hand side */
+    /** Right-hand side of conjunction */
     std::unique_ptr<RamCondition> rhs;
 
 public:
     RamAnd(std::unique_ptr<RamCondition> l, std::unique_ptr<RamCondition> r)
             : RamCondition(RN_And), lhs(std::move(l)), rhs(std::move(r)) {}
 
-    ~RamAnd() override = default;
-
+    /** Get left-hand side of conjunction */
     const RamCondition& getLHS() const {
+        ASSERT(lhs);
         return *lhs;
     }
 
+    /** Get right-hand side of conjunction */
     const RamCondition& getRHS() const {
+        ASSERT(rhs);
         return *rhs;
     }
 
+    /** Print */
     void print(std::ostream& os) const override {
         lhs->print(os);
         os << " and ";
         rhs->print(os);
     }
+
+    /** Get level */
     size_t getLevel() override {
         return std::max(lhs->getLevel(), rhs->getLevel());
     }
 
-    /** Obtains a list of child nodes */
+    /** Obtains list of child nodes */
     std::vector<const RamNode*> getChildNodes() const override {
         return {lhs.get(), rhs.get()};
     }
+
+    /** Create clone */
+    RamAnd* clone() const override {
+        RamAnd* res = new RamAnd(
+                std::unique_ptr<RamCondition>(lhs->clone()), std::unique_ptr<RamCondition>(rhs->clone()));
+        return res;
+    }
+
+    /** Apply */
+    void apply(const RamNodeMapper& map) override {
+        lhs = map(std::move(lhs));
+        rhs = map(std::move(rhs));
+    }
+
+protected:
+    /** Check equality */
+    bool equal(const RamNode& node) const override {
+        assert(dynamic_cast<const RamAnd*>(&node));
+        const RamAnd& other = static_cast<const RamAnd&>(node);
+        return getLHS() == other.getLHS() && getRHS() == other.getRHS();
+    }
 };
 
-/** abstract class for binary operation/relations on values */
+/**
+ * Binary constraint
+ */
+// TODO: rename to RamConstraint
```

Comment:
TODOs should be in the form` // TODO (#issueid): text` or `// TODO (userid): text`

Reply:
Done.